Network policy docs clarifications (#39875)
kubernetes/website: content/en/docs/concepts/services-networking/network-policies.md

Reviewer: If you're willing to update it, the section about `SCTPSupport` is stale.

Author: can you please elaborate?

Reviewer: It's now generally available, no need to enable it and no option to disable it.

Reviewer 2: The whole "SCTP support" section can probably just go away; once you get rid of the stuff about the feature gate, all that's left is a warning that SCTP NetworkPolicies only work if your network plugin supports SCTP, which is kind of a no-brainer.

Author: so do you want me to add another commit to remove SCTP support section? This PR doesn't touch SCTP yet, so maybe it would be a better option to have it as another PR (knowing for how ling we've been discussing this one)?

Author: I added one more commit for SCTP, I left the note about plugin SCTP support, since it looks fine to me in the new "Network traffic filtering" section that talks about protocols. ptal

<!-- overview -->

If you want to control traffic flow at the IP address or port level for TCP, UDP, and SCTP protocols,

> then you might consider using Kubernetes NetworkPolicies for particular applications in your cluster.

Author: I tried to do that under `## Network traffic filtering`, do you think it would be nice to also add a sentence about undefined behaviour for other protocols here in the beginning of the doc?

This ensures that even pods that aren't selected by any other NetworkPolicy will not be allowed
ingress or egress traffic.

## Network traffic filtering

NetworkPolicy is defined for [layer 4](https://en.wikipedia.org/wiki/OSI_model#Layer_4:_Transport_layer) 
connections (TCP, UDP, and optionally SCTP). For all the other protocols, the behaviour may vary 
across network plugins.

{{< note >}}
You must be using a {{< glossary_tooltip text="CNI" term_id="cni" >}} plugin that supports SCTP
protocol NetworkPolicies.
{{< /note >}}

> When a `deny all` network policy is defined, it is only guaranteed to deny TCP, UDP and SCTP

Reviewer: (nit)
```suggestion
When a _deny all_ network policy is defined, it is only guaranteed to deny TCP, UDP and SCTP
```

> connections. For other protocols, such as ARP or ICMP, the behaviour is undefined.

Reviewer: (optionally)
```suggestion
connections. For other protocols, such as ARP or ICMP, the behaviour is undefined: it is up to
each network plugin to decide what behavior to adopt and implement.
```

This ensures that even pods that aren't selected by any other NetworkPolicy will not be allowed
ingress or egress traffic.

> ## L2/L3 behaviour

Reviewer: ```suggestion
## Network traffic filtering
```

NetworkPolicy is defined for L4 connections (TCP, UDP, and optionally SCTP), and for all the
other protocols the behaviour may vary between network plugins.
When `deny all` network policy is defined, it is only guaranteed to deny TCP, UDP and SCTP
connections, for other protocols, like ARP, ICMP, etc. the behaviour is undefined.
Same applies to allow rules: when a specific pod is allowed as ingress source or egress destination,
it is undefined what happens with e.g. ICMP packet, it may be allowed by some network plugins
and denied by others.

> ## SCTP support

Reviewer: Make this a 3rd level heading?
```suggestion
### SCTP support
```

While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
standardized label to target a specific namespace.

## Pod lifecycle

{{< note >}}
The following applies to clusters with a conformant networking plugin and a conformant implementation of
NetworkPolicy.
{{< /note >}}

When a new NetworkPolicy object is created, it may take some for a network plugin
to handle the new object. If the pod that is affected by a NetworkPolicy
is created before the network plugin has completed NetworkPolicy handling,
that pod may be started unprotected, and isolation rules will be applied when 
the NetworkPolicy handling is completed.

Once the NetworkPolicy is handled by a network plugin,

1. All newly created pods affected by a given NetworkPolicy will be isolated before 
they are started.
Implementations of NetworkPolicy must ensure that filtering is effective throughout
the Pod lifecycle, even from the very first instant that any container in that Pod is started.
Because they are applied at Pod level, NetworkPolicies apply equally to init containers,
sidecar containers, and regular containers.

> 2. Allow rules will be applied eventually after the isolation rules (that includes applying both at the same time).

Reviewer 2: "Allow rules will be applied eventually after the isolation rules (or may be applied at the same time)."

In the worst case, a newly created pod may have no network connectivity at all, if
isolation rules were already applied, but no allow rules were applied yet.

Every created NetworkPolicy will be handled by a network plugin eventually, but there is no
way to tell from the Kubernetes API when exactly that happens. 
The only way to check if the network policy is applied to a pod, is to run a connectivity check
from that pod.

The pods must be resilient against being started up with different network

> connectivity than expected. If the pod needs to make sure it has expected network

Reviewer 2: "Therefore, pods must be resilient..."

connectivity before being started, you can use an [init container](/docs/concepts/workloads/pods/init-containers/)

> that will run required connectivity checks before starting the app containers.

Reviewer 2: "...before letting the app containers start." or "...before kubelet starts the app containers".

they are started.
Implementations of NetworkPolicy must ensure that filtering is effective throughout
the Pod lifecycle, even from the very first instant that any container in that Pod is started.
Because they are applied at Pod level, NetworkPolicies apply equally to init containers,
sidecar containers, and regular containers.

2. Allow rules will be applied eventually after the isolation rules (or may be applied at the same time).
In the worst case, a newly created pod may have no network connectivity at all, if
isolation rules were already applied, but no allow rules were applied yet.

Every created NetworkPolicy will be handled by a network plugin eventually, but there is no
way to tell from the Kubernetes API when exactly that happens. 
The only way to check if the network policy is applied to a pod, is to run a connectivity check
from that pod.

Therefore, pods must be resilient against being started up with different network
connectivity than expected. If the pod needs to make sure it has expected network
connectivity before being started, you can use an [init container](/docs/concepts/workloads/pods/init-containers/)

> that will run required connectivity checks before kubelet starts the app containers.

Reviewer 2: As Tim notes, this won't work for checking "negative connectivity", so it might be good to clarify that this is only for the "positive connectivity" case. "If you need to make sure the pod can reach certain destinations before being started, you can use an init container to wait for those destinations to be reachable before kubelet starts the app containers."

Every NetworkPolicy will be applied to all selected pods eventually.
Because the network plugin may implement NetworkPolicy in a distributed manner, 
it is possible that pods may see a slightly inconsistent view of network policies
when the pod is first created, or when pods or policies change.
For example, a newly-created pod that is supposed to be able to reach both Pod A
on Node 1 and Pod B on Node 2 may find that it can reach Pod A immediately,
but cannot reach Pod B until a few seconds later.

## NetworkPolicy and `hostNetwork` pods

> NetworkPolicy behaviour for `hostNetwork` pods is undefined, but should be limited with 2 possibilities:

Reviewer 2: ```suggestion
NetworkPolicy behaviour for `hostNetwork` pods is undefined, but it should be limited to 2 possibilities:
```

> - network plugin can distinguish `hostNetwork` pod traffic from any other traffic and will apply
>   NetworkPolicy to it

Reviewer 2: ```suggestion
- The network plugin can distinguish `hostNetwork` pod traffic from all other traffic
  (including being able to distinguish traffic from different `hostNetwork` pods on
  the same node), and will apply NetworkPolicy to `hostNetwork` pods just like it does
  to pod-network pods.
```

> - network plugin ignores `hostNetwork` pods

Reviewer 2: ```suggestion
- The network plugin cannot properly distinguish `hostNetwork` pod traffic, and so it ignores `hostNetwork` pods when matching `podSelector` and `namespaceSelector`. Traffic to/from `hostNetwork` pods is treated the same as all other traffic to/from the node IP. (This is the most common implementation.)
```

It's not ignoring the pods entirely; eg, a pod which is isolated for ingress will still reject connections from hostNetwork pods.

This applies when
1. a `hostNetwork` pod is selected by `spec.podSelector`.
   
   ```yaml
     ...
     spec:
       podSelector:
         matchLabels:
           role: client
     ...
   ```
 

> 2. a `hostNetwork` pod is selected by a `podSelector` in an `ingress` or `egress` rule.

Reviewer 2: ```suggestion
2. a `hostNetwork` pod is selected by a `podSelector` or `namespaceSelector` in an `ingress` or `egress` rule.
```

While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
standardized label to target a specific namespace.

## Pod lifecycle

When the NetworkPolicy rules should be applied in the pod's lifecycle is undefined, since
there is a potential race if the pod and the NetworkPolicy that selects that pod are created
at around the same time. Nevertheless, if the policy was created a sufficiently long time
before the pod it selects, it is expected that all isolation rules should be set
before any of pod containers are started. Thus, we ensure that a pod can never establish 
an outbound connection or receive an inbound connection that is not allowed by NetworkPolicy 
at any point in its lifecycle. At the same time allow rules may be installed later and pod 

> should be able to handle some delay before its network is fully setup.

Reviewer: Surely it's up to the NetworkPolicy implementation to correct for the desired state no matter which order the two creations happen? I'd rather not document otherwise unless we're explicitly supporting the data race.

Author: Do you mean only leave the first sentence, and let plugins decide everything else?

Reviewer 2: Maybe the text isn't clear: the policy is guaranteed to _eventually_ apply to the pod, it's just not guaranteed to apply to the pod _immediately_.

Basically, we're saying that if you create a NetworkPolicy at time *t* and a Pod at time *t*±*delta* (for small-but-undefined values of *delta*), the network plugin is allowed to process the creations in either order. Given the asynchronous and distributed nature of Kubernetes, this is hard to avoid without adding in performance-destroying serialization/synchronization.

Meanwhile, the last sentence of the new section is making a slightly different point which also maybe isn't clear if you aren't already familiar with the problem. It's saying basically, if the network plugin cannot guarantee that the Pod has _exactly_ the right level of network connectivity when it starts up, then it has to err on the side of being too restrictive. So, pods need to be able to deal with the fact that when their containers first start, they may not yet be able to reach some things that their NetworkPolicies say they're supposed to be able to reach.

Reviewer: Good to know. It wasn't obvious to me that those were the points being made.

I suggest that we frame this to be very clearly about the assumptions a reader can and can't make about their network plugin / NetworkPolicy implementation.

I believe that:
- you _can_ assume that any traffic that you haven't allowed, and that the NetworkPolicy enforces, doesn't arrive at the Pod
  - but only when you have set up cluster networking with support for NetworkPolicy
- you _can't_ assume that the packets allowed by those rules are actually arriving from the moment that the Pod sandbox exists
- you _can't_ assume that the packets that aren't allowed are rejected
- you _can't_ assume that the packets that aren't allowed are dropped
- you _can't_ use the Kubernetes API to determine whether NetworkPolicy is active (actually being enforced)

Author: this is not generally correct, because when the network policy is applied to the pod, you can assume that some packets are allowed or denied for a documented set of protocols, the emphasis of this section is on timing.

as @danwinship said, network policies will be applied eventually. The main outcome for users is that pods should be able to handle some initial delay before these rules are applied. I rephrased the expectation for this timing that hopefully will make more sense for users, please check the updated version

Reviewer: I'd like you to rethink the audience and framing here @npinaeva 

We'll get a tech review for accuracy; from the SIG Docs side, I'm looking to see how well the message will be understood. Right now I'm worried it might not be.

You don't have to use the list-of-assumptions format, but we should make this thought experiment work: 
- show someone this webpage and then ask them about a list of assumptions they can and can't make
- ask them to enumerate the answer (true or false) to the list of assumptions I've stated
- check that someone with existing Kubernetes experience can provide the right answer, given sight of the docs, within say 2 minutes. Ideally, less.

Reviewer 2: > * you _can_ assume that any traffic that you haven't allowed, and that the NetworkPolicy enforces, doesn't arrive at the Pod
>   
>   * but only when you have set up cluster networking with support for NetworkPolicy

correct, subject to the gotcha that this PR is trying to document, which is that this only applies when the NetworkPolicy is created "a sufficiently long time" before the Pod.

> * you _can't_ assume that the packets allowed by those rules are actually arriving from the moment that the Pod sandbox exists

correct

> * you _can't_ assume that the packets that aren't allowed are rejected
> 
> * you _can't_ assume that the packets that aren't allowed are dropped

To be clear, you _can_ assume that packets that aren't allowed will be denied. You just can't assume that they will be denied in any specific way... (Though it's possible that all existing implementations drop?)

> * you _can't_ use the Kubernetes API to determine whether NetworkPolicy is active (actually being enforced)

correct

Reviewer: OK, got it. You can't assume that any traffic that you haven't allowed, and that the NetworkPolicy enforces, doesn't arrive at the Pod - because of eventual consistency for the NetworkPolicy creation.

and there's no `status` / API mechanism to tell when the enforcement is active. We should document that too, as it's a related gotcha.

---

I believe that once the NetworkPolicy is active (which you cannot in general know or observe), new Pods would be covered from the moment they are created and throughout their lifecycle. Have I got that right?

Author: > I believe that once the NetworkPolicy is active (which you cannot in general know or observe), new Pods would be covered from the moment they are created and throughout their lifecycle. Have I got that right?

I would say it is supposed to work like that, but since it was never documented, some plugins may apply network policy rules e.g. some time after the pod is created even if the network policy itself exists for a long time, @danwinship wdyt?

Also the definition of "applied" network policy is  a pain point here. Network policy handles 2 types of pods: pods that are selected by the policy and peer pods that are selected by ingress/egress rules. When we talk about if the network policy is applied to something, we mostly care about the first type. So the network policy is fully applied to a selected pod, when all gress rules together with the default deny are enforced for this pod. 
Since the set of gress pods is constantly changing, it may mean that network policy may never be applied to a specific pod, e.g. if one of the peer pods is always created and deleted, and it never converges. We probably don't want network policy applied condition to be so fragile.
Then we can try to define "applied" part as the deny rules which are more stable, so when the isolation is applied to the pod we say network policy is applied to that pod. Allow rules may be applied/updated independently and will define which endpoint are allowed at any given time, and may be updated with any delay, "eventually"
unfortunately that definition can't be observed from the outside, because if the 2 network policies exist in the namespace, we don't know if the isolation rules are applied because first, second or both network policies are applied.

So this ^ shows that we can't really figure out the status without setting it somewhere in the network policy object.

We can consider introducing a new network policy requirement e.g. with feature gate,to
1. add network policy status "applied" when the isolation rules are enforced on all pods that network plugin knows about at that time (that includes all pods that were Running before the "applied" status was set)
2. all pods selected by the "applied" network policy will be isolated as a part of network setup (i.e. before any containers are started)
That would strictly define that if
1. network policy is "applied" and
2. selected pod is running (actually its network was configured)
then pod is isolated by given network policy.

@danwinship i know something around that was already discussed (probably more than once), so maybe you know if/why ^ is a stupid idea?

Reviewer: We should document how things are now.

If people might infer a guarantee we're not actually making, it's good to make the detail clear. For example, if there can be even a brief moment when a Pod is able to send traffic to another Pod, that's worth knowing because a compromised namespace might deliberately set up conditions to exploit during that window.

This documentation should also allow a cluster operator to look at the behavior of a network plugin and decide whether or not that network plugin properly implements the minimum expectations that we're documenting.

---

We're dropping NetworkPolicy status AIUI. Perhaps we'll be able to emit some metrics or otherwise make it possible to see when a particular NetworkPolicy has reached the first type, applied to Pods in the same namespace that are selected by the NetworkPolicy. However, that's separate to documenting how it works today.

Author: @sftim yeah I agree, it is a completely different topic, but it helps me figure out more accurate words for this section.
I have updated it again, let's see if it explains the current behaviour better

Reviewer 2: > I believe that once the NetworkPolicy is active (which you cannot in general know or observe)

correct

> new Pods would be covered from the moment they are created and throughout their lifecycle. Have I got that right?

No, it's not symmetric: traffic which is denied by the "active" set of NetworkPolicies is guaranteed to be denied from the moment of creation, but traffic which is allowed is not guaranteed to be allowed from the moment of creation.

Eg, if policies say that Pod C should be able to connect to Pod S but not Pod X, then when C comes up, either of these is allowed:

- :smile: When Pod C comes up, it can connect to Pod S, but not Pod X (and this never changes).
- :slightly_smiling_face: When Pod C comes up, it cannot connect to either Pod S or Pod X, but shortly after it comes up, it becomes able to connect to Pod S. (It is never able to connect to Pod X.)

but this is not:

- :angry: When Pod C comes up, it can connect to both Pod S and Pod X, but shortly after it comes up, it stops being able to connect to Pod X.

> I would say it is supposed to work like that, but since it was never documented, some plugins may apply network policy rules e.g. some time after the pod is created even if the network policy itself exists for a long time

It's not even "supposed to work like that". This is distributed computing; the only way we could guarantee that all rules for a pod were fully programmed before the pod comes up would be if we _delayed pod startup_ until all nodes had reported back that they had updated all of their relevant rules. That would be terrible for performance though, so we don't do that.

> @danwinship i know something around that was already discussed (probably more than once), so maybe you know if/why ^ is a stupid idea?

There are too many moving parts. You can't easily say that a NetworkPolicy is now "active"/"being enforced", because it may not happen at the same time on all nodes. And knowing that a NetworkPolicy is "active" doesn't let you know if it's being applied to a particular pod yet or not in all cases. (eg, there's also the case where the labels on a pod change, which changes the set of policies that apply to it.)

Network plugins may implement extra `status.conditions` for pod readiness based on the
status of network policies. These conditions then may be used by 
[Pod readiness gate](/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate).

> ## `hostNetwork` pods

Author: I didn't include `NetworkPolicy and` part, since the whole document is about network policies, so I am not sure if it adds value. 
Second-level heading would be nice, but I guess we need to find a set of headers that would go under it, otherwise it is difficult to figure out the right heading name if there will be only 1 sub-section. Maybe you have an idea of what else could go under a new second-level header?

> ## `hostNetwork` pods

Author: I am not sure I understand how exactly that is supposed to work, and I since this PR already covers 3 new sections, maybe we better document that limitation in another PR?

> ## `hostNetwork` pods

Author: sure I am happy to improve it, but I can't figure out a better name. Updated it to "NetworkPolicy and `hostNetwork` pods" for now.

NetworkPolicy behaviour for `hostNetwork` pods is undefined, when
1. `hostNetwork` pod is selected by `spec.podSelector`.
For example, if there is a `hostNetwork` pod on the `nodeA`, that has label `role: client`,
the following NetworkPolicy:
```yaml
  ...
  spec:
    podSelector:
      matchLabels:
        role: client
  ...
```
may apply ingress and egress rules to all connections with `nodeA` ip, or just ignore

> that pod as if it didn't match the selector, or do something else.

Reviewer: Please indent all this text by 3 characters; it makes the ordered list more obvious.

2. `hostNetwork` pod is selected by `ingress` or `egress` rules.
For example, if there is a `hostNetwork` pod on the `nodeA`, that has label `role: client`,
the following NetworkPolicy:
```yaml
  ...
  ingress:
    - from:
      - podSelector:
          matchLabels:
            role: client
  ...
```
may allow all ingress connections with `nodeA` ip, or just ignore that pod as 
if it didn't match the selector, or do something else.

> At the same time, since `hostNetwork` pods have the same ips as the nodes they reside on,

Reviewer: ```suggestion
At the same time, since `hostNetwork` pods have the same IP addresses as the nodes they reside on,
```

While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
standardized label to target a specific namespace.

## Pod lifecycle

> When the NetworkPolicy rules should be applied in the pod's lifecycle is undefined, since

Reviewer 3: This sentence is not good English, "when ..., since ....".

there is a potential race if the pod and the NetworkPolicy that selects that pod are created

> at around the same time. Nevertheless, if the policy was created a sufficiently long time

Reviewer 3: Why there needs a "sufficiently long time"?
How do we define "sufficiently long"?

Author: I removed that part, and converted it to the less-detailed "initial intent" explanation

before the pod it selects, it is expected that all isolation rules should be set
before any of pod containers are started. Thus, we ensure that a pod can never establish 
an outbound connection or receive an inbound connection that is not allowed by NetworkPolicy 
at any point in its lifecycle. At the same time allow rules may be installed later and pod 
should be able to handle some delay before its network is fully setup.

Network plugins may implement extra `status.conditions` for pod readiness based on the
status of network policies. These conditions then may be used by 
[Pod readiness gate](/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate).

## `hostNetwork` pods

NetworkPolicy behaviour for `hostNetwork` pods is undefined, when

> 1. `hostNetwork` pod is selected by `spec.podSelector`.

Reviewer 3: ```suggestion
1. a `hostNetwork` pod is selected by `spec.podSelector`.
```

>    For example, if there is a `hostNetwork` pod on the `nodeA`, that has label `role: client`,

Reviewer 3: remove quotes around nodeA.

   the following NetworkPolicy:
   ```yaml
     ...
     spec:
       podSelector:
         matchLabels:
           role: client
     ...
   ```

>    may apply ingress and egress rules to all connections with `nodeA` ip, or just ignore

Reviewer 3: You may want to emphasize that the Pod IP, in this case, is the node IP.

nit: spell 'IP' rather than 'ip'.

   that pod as if it didn't match the selector, or do something else.

> 2. `hostNetwork` pod is selected by `ingress` or `egress` rules.

Reviewer 3: ```suggestion
2. a `hostNetwork` pod is selected by `ingress` or `egress` rules.
```

While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
standardized label to target a specific namespace.

## Pod lifecycle

We say that a NetworkPolicy is applied to a selected pod, when isolation rules are enforced
for that pod. At that point pod can never establish an outbound connection or receive
an inbound connection that is not allowed by the NetworkPolicy. Allow rules will be applied
eventually after the isolation rules (that includes applying both at the same time).
That means, that once the NetworkPolicy is applied to a pod, the pod should be able to handle situations,
when it can't establish a connection that is allowed by an existing NetworkPolicy for some time.
The requirements for Allow rules are less strict, because a set of allowed endpoints may be constantly
changing if the rules are based on a `namespaceSelector` or a `podSelector`.

Every NetworkPolicy will be applied to all selected pods eventually, but there is no
way to tell from the existing Kubernetes API when exactly the network policy is applied to a specific

> pod. The only to check if the network policy is applied to a pod, is to run a connectivity check.

Reviewer: ```suggestion
pod. The only way to check if the network policy is applied to a pod, is to run a connectivity check.
```

Even that connectivity check isn't 100% reliable, IMO.

Author: I was trying to specify at least 1 way to check if the pod is isolated or not. I can just remove it, or replace with some better option? @danwinship do we want to answer the question "how I am supposed to figure out if the network policy is applied to a pod or not"?

While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
standardized label to target a specific namespace.

> ## Pod lifecycle

Reviewer: ##

While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
standardized label to target a specific namespace.

## Pod lifecycle

A NetworkPolicy is applied to a selected pod, when isolation rules are enforced
for that pod. At that point pod can never establish an outbound connection or receive
an inbound connection that is not allowed by the NetworkPolicy. Allow rules will be applied
eventually after the isolation rules (that includes applying both at the same time).
That means, that once the NetworkPolicy is applied to a pod, the pod should be able to handle situations,

> when it can't establish a connection that is allowed by an existing NetworkPolicy for some time.

Reviewer 2: hm... this is all correct, but it's not very clear.

My attempt:

Because the network plugin may implement NetworkPolicy in a distributed manner (for example, with each node enforcing the policies only for pods on that node), it is possible that pods may see a slightly inconsistent view of network policies when the pod is first created, or when pods or policies change. For example, a newly-created pod that is supposed to be able to reach both Pod A on Node 1 and Pod B on Node 2 may find that it can reach Pod A immediately, but cannot reach Pod B until a few seconds later.

Network plugins should attempt to ensure that all policy rules are correctly configured as soon as possible after pod creation, but pods must be resilient against being started up with less network connectivity than expected (or in the worst case, no network connectivity at all, since if the network plugin can't guarantee that the rules are exactly right at pod startup, it must err on the side of making them too restrictive, to ensure that the pod can never establish an outbound connection or receive an inbound connection that is not allowed by NetworkPolicy).

Every NetworkPolicy will be applied to all selected pods eventually, but there is no
way to tell from the existing Kubernetes API when exactly the network policy is applied to a specific

> pod. The only way to check if the network policy is applied to a pod, is to run a connectivity check.

Reviewer 2: I'd remove the word "existing" since the discussions around NetworkPolicyStatus led to a consensus that there is no good way to ever do this from the Kubernetes API, particularly given the "distributed" thing where different pods may have different perceptions of the current status.

As for whether "connectivity check" is reliable; it's reliable if and only if you do the connectivity check _from that pod_. So maybe we could suggest that if a pod can't be rewritten to be resilient to lack-of-connectivity, that you could add an `initContainer` to it that checks that it has the connectivity it needs, so that things will be working as expected when the main containers start.

Network plugins may implement extra `status.conditions` for pod readiness based on the
status of network policies. These conditions then may be used by 

> [Pod readiness gate](/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate).

Reviewer 2: This document is written for end users, whereas this comment seems to be targeted more at network plugin authors? We probably shouldn't be suggesting to end users that they can do this if we don't know that there are already network plugins that implement that...

## NetworkPolicy and `hostNetwork` pods

> NetworkPolicy behaviour for `hostNetwork` pods is undefined, when

Reviewer 2: "undefined" makes it sound like this is an unanswered question, but AFAIK literally every network plugin implements this the same way: a `podSelector` (whether spec or gress) never selects `hostNetwork` pods.

(In theory, Cilium (or any other plugin using at least some eBPF) could do something better here, by applying some rules based on cgroups rather than only using IP addresses. But AFAIK nobody actually currently does this, so it seems like we should just officially bless the current behavior and revisit this in NPv2.)

Author: I may know one plugin that interprets peer hostNetwork pods as nodeIP, since it is the only way (I can think of) to allow kubeapi pods without hardcoding master node ips.
I am not sure if we can say that hostNetwork pods from the `spec` are always ignored, just because we don't know of any plugins that do something else. Maybe we better say something like "most network plugins" will ignore, but it is still officially undefined?

1. a `hostNetwork` pod is selected by `spec.podSelector`.
   
   For example, if there is a `hostNetwork` pod on the nodeA (pod IP equals nodeA IP in that case), 
   that has label `role: client`, the following NetworkPolicy:
   ```yaml
     ...
     spec:
       podSelector:
         matchLabels:
           role: client
     ...
   ```
   may apply ingress and egress rules to all connections with nodeA IP, or just ignore

>    that pod as if it didn't match the selector, or do something else.

Reviewer 2: no, it is unambiguously wrong to apply the networkpolicy to nodeA IP traffic that _isn't_ to/from the hostNetwork pod. You have to either apply the policy _only_ to the hostNetwork traffic, or not apply it to any nodeA IP traffic

Author: I tried to just explain what undefined behaviours may be, since i guess we agreed on it here https://github.com/kubernetes/website/pull/39875#discussion_r1168498073
I can add one more sentence saying that most plugins will ignore hostNetwork pods, or do you prefer some other option?

Reviewer 2: "Undefined" here doesn't mean that the plugin can do anything it wants. There are only two permitted possibilities:
1. The plugin implements the policy "correctly", by applying it to traffic to/from the hostNetwork pod, but _not_ applying it to other traffic with the node IP. Most plugins are not able to do this though.
2. The plugin ignores `hostNetwork` pods for purposes of matching `podSelector`s. (This is the expected behavior, though that's not documented.)

The plugin _can't_ "apply ingress and egress rules to all connections with nodeA IP". That's just wrong.

While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
standardized label to target a specific namespace.

## Pod lifecycle

When a new NetworkPolicy object is created, it may take some for a network plugin
to handle the new object. If the pod that is affected by a NetworkPolicy
is created before the network plugin has completed NetworkPolicy handling,
that pod may be started unprotected, and isolation rules will be applied when 
the NetworkPolicy handling is completed.

> Once the NetworkPolicy is handled by a network plugin,

Reviewer: ```suggestion
For a cluster with a conformant networking plugin and a conformant implementation of
NetworkPolicy:
```

Reviewer 2: (IMO we don't need this note; the entire page only applies to plugins that implement NetworkPolicy, and the entire `/services-networking/` section only (fully) applies to clusters with conformant networking plugins.)

Author: yeah good point, @sftim what do you think?

Reviewer: > the entire page only applies to plugins that implement NetworkPolicy

The conceptual explanation is relevant even if your cluster is using a network plugin that has never heard of NetworkPolicy.

---

We should document _somewhere_ what we expect from a conforming network plugin. Otherwise, it's not fair on plugin authors - they need to know what's expected and shouldn't have to become detectives in meeting that end.

The way we inform implementers might once day be through a separate reference page, but for now this concept page is the closest thing we have.

> 1. All newly created pods affected by a given NetworkPolicy will be isolated before 

Reviewer 4: > pods...will be isolated before they are started

I believe this bullet point is incorrect. This requirement does not even seem possible for a NetworkPolicy implementation which is disjoint from the CNI (e.g. Azure NPM).

Assuming that isolation here means blocking traffic. Please correct me if I'm misinterpreting.

Reviewer 2: Your interpretation, and the bullet point, are correct. A correct implementation of NetworkPolicy must ensure that pods are not able to bypass pre-existing NetworkPolicies at any point in their lifecycle.

If pods aren't fully subject to NetworkPolicy at startup time, then an attacker can exploit that race condition by continuously creating new pods and trying to make connections that should have been blocked, until they eventually succeed, and then they can do whatever bad thing the NetworkPolicy was supposed to have been preventing them from doing, rendering the NetworkPolicy useless.

> This requirement does not even seem possible for a NetworkPolicy implementation which is disjoint from the CNI (e.g. Azure NPM).

Note that it's allowed for pods to come up _more restricted_ than they're supposed to be, just not _less restricted_. So if the NP impl can create a fallback "block everything else" rule that would apply to newly-created pods, then it could do that, and then the pod would come up fully isolated until the NP impl has a chance to fix it up.

If the CNI plugin works in such a way that the NP implementation can't create rules that would apply to a new pod before that pod exists (eg, the CNI plugin is creating a new interface), then there is probably no way to correctly implement NetworkPolicy independently of the CNI plugin. (Though there are small-ish modifications that could be made to the plugin to make this possible. Eg, you could add a flag to the CNI plugin saying "make newly-created pods start out completely isolated", and then it would depend on the existence of some external NP impl to fix the pods up after that.)

Reviewer: > make newly-created pods start out completely isolated

_(aside)_
That sounds like a very reasonable workaround, especially given the (alpha) `PodReadyToStartContainers` [condition](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) - the new Pod might be fully isolated, then add packet filtering rules, _then_ remove the full isolation, and finally assert `PodReadyToStartContainers`. But: it does need the network plugin to play at least some role.

Maybe there's an interface or API that we can define so that a network plugin waits between setting up IP networking for a Pod and asserting that condition. That would help standalone NetworkPolicy implementations to provide the guarantees we expect.

> they are started.

Reviewer: ```suggestion
they are started.
Implementations of NetworkPolicy must ensure that filtering is effective throughout
the Pod lifecycle, even from the very first instant that any container in that Pod is started.
Because they are applied at Pod level, NetworkPolicies apply equally to init containers,
sidecar containers, and regular containers.
```

2. Allow rules will be applied eventually after the isolation rules (that includes applying both at the same time).
In the worst case, a newly created pod may have no network connectivity at all, if
isolation rules were already applied, but no allow rules were applied yet.

> Every created NetworkPolicy will be handled by a network plugin eventually, but there is no
> way to tell from the Kubernetes API when exactly that happens. 

Reviewer: ```suggestion
Every created NetworkPolicy will be handled by a network plugin eventually, but there is no
way to tell from the Kubernetes API when exactly the policy is being enforced for new or
existing Pods.
(As an example: if you write a NetworkPolicy implementation that begins enforcing the policy from
midnight the next day, that conforms to Kubernetes' requirements - but of course could be very
frustrating for cluster administrators).
```

Author: This suggestion looks more implementor-focused to me, and I think we were going to outline implementor guideline in a different document, wdyt?
Another note - I don't think the implementation you described will pass conformance tests for network policy, so maybe it is not really conformant?

Reviewer 2: There are no _conformance_ tests for NetworkPolicy, but it wouldn't pass the ordinary e2e tests.

> I think we were going to outline implementor guideline in a different document

yeah, there's been discussion about that (partly inspired by this PR). https://groups.google.com/g/kubernetes-sig-network/c/Ga9SWGs00k4 / https://github.com/kubernetes/community/issues/7421

The only way to check if the network policy is applied to a pod, is to run a connectivity check

> from that pod.

Reviewer: Even then, you might not be able to tell (I think).

Author: that is right, e.g. connections may be dropped for another reason, but I am not sure if adding these details will improve or over-complicate this section, wdyt?

Reviewer: I'd omit the sentence. I'm wary about the risk of misleading people.

(I don't want someone to think “_Right, so this Pod shouldn't be able to talk to a Pod in namespace `untrusted`. I tried and failed, so my policy is working. Job done, :shipit:!_”)

standardized label to target a specific namespace.

## Pod lifecycle

1. Isolation rules are applied before allow rules

   A NetworkPolicy is applied to a selected pod, when isolation rules are enforced
for that pod. At that point pod can never establish an outbound connection or receive
an inbound connection that is not allowed by the NetworkPolicy.
Allow rules will be applied eventually after the isolation rules (that includes applying both at the same time).

   In the worst case, a newly created pod may have no network connectivity at all, if 
isolation rules were already applied, but no allow rules were applied yet. 

2. NetworkPolicy is not guaranteed to be applied to a pod before it is started

   Network plugins should attempt to ensure that all policy rules are correctly
configured as soon as possible after pod creation, but how soon that actually
happens may depend on the network plugin implementation, current cluster load, 
Pod and NetworkPolicy creation order, etc.

   If the NetworkPolicy is not applied to a pod before it is started,
no isolation rules may be present for some time, and pod may be able to establish

> a connection that is denied by an existing NetworkPolicy.

Reviewer 2: This makes it sound like a network plugin might just _choose_ to not apply NetworkPolicies, which isn't what we want to be saying at all. I think maybe you're trying to use "applied" as a term of art, defined in the previous section, but it doesn't really come across that way.

I've generally tried to express this in terms of the network plugin "being aware" / "becoming aware" of the NetworkPolicy. So I'd say:

1. There's no way to know for sure when the network plugin becomes aware of a newly-created NetworkPolicy, and depending on how the plugin is implemented, it may be aware of the policy on some nodes but not others.
2. Once the network plugin is aware of the policy (on a node), any pod (on that node) selected by that policy is guaranteed to be isolated.
3. Every NetworkPolicy will be fully applied to all selected pods eventually, ...

Reviewer 2: OK, I get not wanting to be too plugin-focused, but I don't think the "applied to" language is very clear. Actually, I think a lot of what makes it unclear is that you're not talking about _who_ is doing the applying, so it's easy to interpret as "when the user applies a NetworkPolicy to a pod" etc.

IMO if we are talking about it from the user perspective then the relevant points are:

- If you create a pod when there are pre-existing NetworkPolicies that select that pod, then it is guaranteed that the pod will be "isolated" when it starts, and the pod will never have more network connectivity than it is supposed to. However, in some cases it may take a moment after the pod starts for all of the relevant "allow" rules to come into effect.
- If you create a pod and a NetworkPolicy that selects that pod at approximately the same time, and there were not any pre-existing NetworkPolicies that selected that pod, then it is not guaranteed that the pod will be isolated when it starts, because (regardless of which order you created the two objects in), it might take Kubernetes longer to process the NetworkPolicy creation than the Pod creation.

From the end user's perspective, that's all you need to know.

Reviewer: > If you create a pod when there are pre-existing NetworkPolicies that select that pod, then it is guaranteed that the pod will be "isolated" when it starts, and the pod will never have more network connectivity than it is supposed to. However, in some cases it may take a moment after the pod starts for all of the relevant "allow" rules to come into effect.

This only holds if you have installed a compatible network plugin. Plenty of clusters use a network plugin where that statment doesn't hold. Specifically, these parts would be false:

Reviewer: End users _do_ need to know this caveat!

Reviewer 2: Right, that part is important too but I think it probably belongs earlier on in the docs. If you get this far, we're assuming you actually have working NP, I think

Author: > If you create a pod when there are pre-existing NetworkPolicies that select that pod

I think we've discussed it before that it is impossible to say which policies are "pre-existing", more specifically it means that the policies are created some time before the pod is created, and then we can't define how much time it should be exactly. While there is no status on network policy from the plugin saying that this policy was handled, we don't know when that happens.

That is generally why this section is so complicated.

Author: I restructured this paragraph to focus more on what is guaranteed after the network policy is handled by the network plugin, @danwinship lmk what you think about this version

on Node 1 and Pod B on Node 2 may find that it can reach Pod A immediately,
but cannot reach Pod B until a few seconds later.

## NetworkPolicy and `hostNetwork` pods

NetworkPolicy behaviour for `hostNetwork` pods is undefined, when
1. a `hostNetwork` pod is selected by `spec.podSelector`.
   
   For example, if there is a `hostNetwork` pod on the nodeA (pod IP equals nodeA IP in that case), 
   that has label `role: client`, the following NetworkPolicy:
   ```yaml
     ...
     spec:
       podSelector:
         matchLabels:
           role: client
     ...
   ```
   may apply ingress and egress rules to all connections with nodeA IP, or just ignore
   that pod as if it didn't match the selector, or do something else.

> 2. a `hostNetwork` pod is selected by `ingress` or `egress` rules.

Reviewer 2: ```
2. a `hostNetwork` pod is selected by a `podSelector` in an `ingress` or `egress` rule.
```
(since there's the `ipBlock` case too, but you don't talk about that in this bullet point)
